File: engine/src/main/java/org/qubership/integration/platform/engine/service/testing/TestingContext.java
package org.qubership.integration.platform.engine.service.testing;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.charset.StandardCharsets;
import java.util.Base64;

/**
 * Identifies a mocked call for the testing service. The header name, the field names and the standard, padded
 * base64 alphabet are a contract with the Go service, which parses {@code path} and {@code operationPath} to
 * feed its query- and path-parameter matchers.
 */
@JsonPropertyOrder({"chainId", "elementId", "operationPath", "path"})
public record TestingContext(String chainId, String elementId, String operationPath, String path) {

    public static final String HEADER_NAME = "Testing-Service-Context";

    private static final ObjectMapper MAPPER = new ObjectMapper();

    public String encode() {
        try {
            byte[] json = MAPPER.writeValueAsString(this).getBytes(StandardCharsets.UTF_8);
            return Base64.getEncoder().encodeToString(json);
        } catch (JsonProcessingException exception) {
            throw new IllegalStateException("Failed to encode the testing service context", exception);
        }
    }
}

File: testing-service/internal/model/testing_context.go
package model

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
)

// TestingContextHeader names the header that carries a base64-encoded
// TestingContext. Both the header name and the field names below are part of
// the contract with the engine and must not be renamed.
const TestingContextHeader = "Testing-Service-Context"

type TestingContext struct {
	ChainID       string `json:"chainId"`
	ElementID     string `json:"elementId"`
	OperationPath string `json:"operationPath"`
	Path          string `json:"path"`
}

// DecodeTestingContext reads the value of TestingContextHeader.
func DecodeTestingContext(encoded string) (*TestingContext, error) {
	data, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return nil, fmt.Errorf("decode %s header: %w", TestingContextHeader, err)
	}
	var testingContext TestingContext
	if err := json.Unmarshal(data, &testingContext); err != nil {
		return nil, fmt.Errorf("parse %s header: %w", TestingContextHeader, err)
	}
	return &testingContext, nil
}
